### PR TITLE
WASM: Blocks signatures with multi-results

### PR DESCRIPTION
**Pull request by Cellule**
WebAssembly: Support signature in blocks with in parameters and multi-results. 
Ground work for functions with multi-results.

Related: #3478

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/4712)

### CODE REVIEW

File: lib/WasmReader/WasmSignature.cpp [hunk 1]
```diff
@@ -22,6 +21,10 @@ WasmSignature::WasmSignature() :
 
 void WasmSignature::AllocateParams(Js::ArgSlot count, Recycler * recycler)
 {
+    if (GetParamCount() != 0)
+    {
+        throw WasmCompilationException(_u("Invalid call to WasmSignature::AllocateResults: Results are already allocated"));
```
Comment (Penguinwizzard) [inline lib/WasmReader/WasmSignature.cpp]: nit: s/Results/Params/g?

File: lib/WasmReader/WasmSignature.cpp [hunk 2]
```diff
@@ -206,24 +240,54 @@ uint32 WasmSignature::WriteSignatureToString(_Out_writes_(maxlen) char16* out, u
...
+        WRITE_BUF(_u(")->void"));
+    }
+    else if (GetResultCount() == 1)
+    {
+        WRITE_BUF(_u(")->%ls"), WasmTypes::GetTypeName(this->GetResult(0)));
+    }
+    else
+    {
+        WRITE_BUF(_u(")->("));
+        for (uint32 i = 0; i < GetResultCount(); i++)
```
Comment (Penguinwizzard) [inline lib/WasmReader/WasmSignature.cpp]: nit: You may want to add handling for maxlen here and report ")->(…)" or something like that if there's not enough space.

File: test/WasmSpec.MultiValue/simple_block.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  // so there's no good way to test for specific bitpatterns here.
+  let actual = action();
+  if (!Number.isNaN(actual)) {
+    throw new Error("Wasm return value NaN expected, got " + actual);
+  };
+}
+
+
+// simple_block.wast:6
+let $1 = instance("\x00\x61\x73\x6d\x01\x00\x00\x00\x01\x8b\x80\x80\x80\x00\x02\x60\x00\x01\x7f\x60\x02\x7f\x7f\x01\x7f\x03\x82\x80\x80\x80\x00\x01\x00\x07\x8a\x80\x80\x80\x00\x01\x06\x70\x61\x72\x61\x6d\x73\x00\x00\x0a\x90\x80\x80\x80\x00\x01\x8a\x80\x80\x80\x00\x00\x41\x01\x41\x02\x02\x01\x6a\x0b\x0b");
```
Comment (Penguinwizzard) [inline test/WasmSpec.MultiValue/simple_block.js]: Is there a way that we can generate this through a wabt call instead of hard-coding the wast?

Comment (Cellule) [reply test/WasmSpec.MultiValue/simple_block.js]: Currently there is no support in wabt for multi-value, which is why I used the spec interpreter to generate this file.
It's not ideal, but it should be a temporary solution until there is support in wabt.

File: lib/WasmReader/WasmSignature.cpp (hunk 1 above)
Comment (Cellule) [reply lib/WasmReader/WasmSignature.cpp]: fixed

File: lib/WasmReader/WasmSignature.cpp (hunk 2 above)
Comment (Cellule) [reply lib/WasmReader/WasmSignature.cpp]: done

**Comment by Cellule** [comment]
The discussion in https://github.com/WebAssembly/multi-value/issues/3 made me realize we didn't read the wasm Language types properly making the extension of blocktype confusing.
I took the opportunity to fix other instances of this issue as well.

**Pull request merged**